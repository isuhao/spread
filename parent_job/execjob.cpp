#include "execjob.hpp"
#include <assert.h>
#include <stdexcept>

using namespace Spread;

bool ExecJob::execJob(JobPtr p, bool failOnError)
{
  assert(!jobs.size());
  jobs.insert(p);

  lastJob = p->getInfo();
  runClient(p, true, false);
  jobs.erase(p);
  done.push_back(p);

  if(failOnError && lastJob->isNonSuccess())
    {
      if(lastJob->isError()) throw std::runtime_error(lastJob->getMessage());
      else throw std::runtime_error("Client job aborted: " + lastJob->getMessage());
    }

  return lastJob->isSuccess();
}
