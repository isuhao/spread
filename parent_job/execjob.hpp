#ifndef __SPREAD_EXEC_JOB_HPP_
#define __SPREAD_EXEC_JOB_HPP_

#include "parentjob.hpp"

/* ExecJob is a ParentJob that has at most one active child at any
   given time.

   Rather than running the child in a new thread, the child is
   executed in the same thread as the ExecJob itself.
 */

namespace Spread
{
  struct ExecJob : ParentJob
  {
  protected:
    JobInfoPtr lastJob;

    /* Execute 'p' as a child of this job. The job is executed
       in-thread. The child lists are set up accordingly, making the
       child visible to the outside. Throws on error.

       If failOnError=false, child jobs are allowed to fail. On
       non-success, execJob() will then return false. You may inspect
       lastJob for details.
     */
    bool execJob(JobPtr p, bool failOnError=true);

    /* Convenience pointer version. Must be called with a newly
       created object that is not yet owned by a shared_ptr.
    */
    bool execJob(Job *p, bool failOnError=true)
    { return execJob(JobPtr(p), failOnError); }
  };
}
#endif
