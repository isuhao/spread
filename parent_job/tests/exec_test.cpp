#include <iostream>
#include "execjob.hpp"
//#include "common.cpp"

using namespace std;
using namespace Spread;

/*
  Things to test:

  - that both versions of execJob work
  - that lastJob is set correctly
  - that client success, errors, exceptions and aborts work as they
    should
  - that abort() requests are propagated to the client
  - that jobs are put in the correct list at all times
  - that we can continue running after a failed job, but also
    intercept it and choose to exit
  - that progress and status messages work
 */

int main()
{
  cout << "Hello\n";

  return 0;
}
