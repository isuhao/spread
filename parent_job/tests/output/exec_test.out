Staring main job
start 0
Testing bwait
Returned from bwait
start 2
end 2
start 1
Taking a break
Back in MAIN, while the job is taking a break!
BUS: LUNCH BREAK 3/4
  DONE:
  SUC: Success 0/0
  JOBS:
  BUS: LUNCH BREAK 3/4
Break's over, time to work
end 1
Got: LUNCH BREAK
start 3
Got expected abort
start 4
Got expected error: I'm allergic to the number 4
Next job will abort us.
start 5
MAIN: Aborting caller thread now.
BUS:  0/0
  DONE:
  SUC: Success 0/0
  SUC: Success 3/4
  ABO:  0/0
  ERR: I'm allergic to the number 4 5/6
  JOBS:
  BUS:  0/0
Ok, returning.
Got ABORT in i=5 thread, exiting.
Got ABORT in main job, exiting.
ABO:  0/0
  DONE:
  SUC: Success 0/0
  SUC: Success 3/4
  ABO:  0/0
  ERR: I'm allergic to the number 4 5/6
  ABO:  0/0
EXIT
